### DISCUSSION TITLE
Compile core files AoT and remove builder references from core

### DISCUSSION THREAD

**Discussion by edusperoni** [NativeScript/rfcs#8] [1. Ideas & Discussions]
Nativescript's core flavor isn't tree-shakeable due to it being JIT compiled, which requires fully fledged parsers and the UI Builder to be bundled with every Nativescript app as they're used by many plugins and even core functions as well.

**Proposal:** create a xml loader for Nativescript's core flavor so xml files will be converted into their "code-behind" counterparts.

All Builder references from core could then be removed/changed so the functions now would only require a `View` or `() => View`, instead of a `ViewEntry`.

This goes hand in hand with #4 

*Possible breaking changes*:
it may be possible to analyze for usages of `ViewEntry` and replace them with the proper imports, but we may need to do something like:

```typescript
import modalPage from './modal-page.xml';
```

to use in showing of modals and other dynamically created pages.

**Comment by rigor789** [comment]
I was going to open a similar issue, but mainly about extracting core flavor into it's own package.

**Comment by farfromrefug** [comment]
I would love to see that extracted out in a package. Other flavors don't need this

**Comment by edusperoni** [comment]
That was going to be my suggestion, but as @NathanaelA said in a TSC meeting, many plugins use the core flavor for custom views.

I think it all boils down to how big of a breaking change we want to introduce

If we do want to move forward with removing the core flavor, I think we'll be able to clean up a good chunk of functions from core, like `bindingContext`, which is very confusing and is even used in things like ListView as a way to select the template (it creates a Label with the template selector function bound to a property and updates the bindingcontext to get the new template key).

**Comment by farfromrefug** [comment]
When you say plugins depend on core flavor can you elaborate? Not sure I see the issue.
If I take my newly created drawer plugin, it depends on the core Ui in the sense that the drawer extends GridLayout. But it does not depend on the XML/ builder part of it. 
Maybe you are talking more the listview kind of plugins where you can have references to the builder to "inflate" cell views?
I think we could easily split in different packages for a major version and have plugins updated to either really "split" the work or at least use the separated plugins.
I think with the efforts that have been put in federating the plugins it would be much easier to quickly have plugins updated to such a major version change

**Comment by edusperoni** [comment]
No, I mean some plugins actually have a xml file with their layouts and inflate them when needed. This is just something to consider when we introduce this breaking change.

**Comment by farfromrefug** [comment]
@edusperoni ok ! have not seen that anywhere but you are right, have to be considered

**Comment by rigor789** [comment]
Would be useful to gather a few plugins that do this - and perhaps we can handle them in an xml-loader without requiring plugins to change. Webpack would compile the xml and it would be passed to the same runtime methods these plugins use, but handled seamlessly. (this definitely needs thorough testing - but should be doable - unless the xml is generated from for eg. string literals)

**Comment by NathanWalker** [comment]
Here’s a plugin that we use often that uses string literal:
https://github.com/davecoffin/nativescript-filterable-listpicker/blob/34cbf49b13adf2ff831f98330aa4d45fc4d0d7ee/src/filterable-listpicker.ts#L89

**Comment by NathanaelA** [comment]
1.  Many plugins use String literal templates; and some even use a XML template if it is more complicated.     Just use searchcode to do a search...    I think I've even seen some vue code that used pure core templates for items that had actual sub-templates.

2. Small side topic `bindingContext` is used by all flavors and in all plugins, to my knowledge under the hood.  That would have to remain.

I think this is probably a unrealistic goal, as too many things depend on it and it would create tremendous chaos in the community as now a huge number of plugins have to be upgraded to "import" core/vanilla for them to work.     My 2 cents...  :grinning: 

However, I would be in support of building something that automagically converts the xml into native code as part of the bundler; while leaving the builder in place to handle any dynamic uses of it...

**Comment by edusperoni** [comment]
I believe we can probably do some parsing and try to compile AoT those string literals. Things would probably get hairy with template strings, though.

Regarding `bindingContext`, I know Angular have no references to it, and searching for it in Vue/React/Svelte yielded no results as well. Most of the references are on core, more especifically:

`Binding`, which seems to be the "meat" of the binding context implementation, and `ViewBase`, where it wires up `Binding`, and some random places like `ListView` where it uses `bindingContext` to evaluate a function if it was passed as a string (which shouldn't really happen, as the binding should be able to do this conversion before passing it to the View).

This **could** be all extracted from core into a "vanilla" flavor which would do this wiring when the view requires it. Yes, it would also be a big breaking change and plugins would likely need to be updated, which is the same that happened with v7.0 when all package names changed. In fact, @NathanWalker's example still isn't compatible with 7.0.

All flavors seem to be able to convert their components' inputs into valid core inputs, it would be really nice if "vanilla" could do the same, even if it were still in the same package as core. I'm working on removing things like `zonedCallback` because I think flavor specific stuff shouldn't really "leak" into applications that don't need it and make core harder to maintain.

**Comment by NathanaelA** [comment]
@edusperoni - 
Sorry on `bindingContext` I was thinking `Observable`.  LOL, however since we (or me :grinning:) are now both talking about the same thing... 

- BindingContext is used as part of the result set in ListView (and other paging items like pager, repeater, etc).   For example if you put a tap event on a listview, the event will give you a bindingContext of the data, now of course this "bindingContext" may or may not be an observable, depending on what you pass into the component..
- bindingContext is also used for allowing items to be bound to screen during building it.  (i.e. Page.bindingContext).
- certain built in major events (loaded, navigation) will give you the bindingContext (which that data might not be relevant to the other flavors; but might not be worth trying to abstract them out and creating extra additional layers)
- Svelte is only truly a minor wrapper around core; which means virtually everything in core is also used in the Svelte flavor.  
- Around 200 plugins with "bindingContext" references, although many of those could be the demo, don't really want to look at ~2000 separate files to see for sure...  :grinning:

I'm not sure it can easily be removed from the core without creating other issues, but I'm not opposed to it.  :grinning:


> I'm working on removing things like zonedCallback
:+1:

Builder & String Literals; we could technically have any plugin that needs "builder" include it, to eliminate any AOT issues...
But in all reality I'm not sure removing Builder (~30k) adds/removes that much for core to make it worth the time investing to get rid of it out of core.   The overhaul to remove it probably isn't worth the time for the very small amount of code it actually is.      It would be much better use of time to figure out how to make Builder (& other flavors) dynamically include the components they are using, rather than bundling everything in.  All the components included by default adds the majority of the overhead. (~8 meg)    Maybe we need to have a "register" function for core, so that when you import("ui/Label") it registers it as part of the file, so then all the ui components are able to be tree shaken, where right now core registers all of them so everything is included in every flavor.   

Brainstorming ideas along this line; maybe each flavor could "override" the a register handler function so they can setup what they need to during the registration.   Maybe what happens is the current global.registerElement function checks to see if (global.__RegisterComponent) is setup, if not it creates an cache of anything registered.  Once a flavor is imported; it sets the `global.__RegisterComponent  = Vue.registerFunction` (or whatever the flavors registerElement is called) and then passes a asked alsoo does a global.registerElement(null) to force the registerElement handle cache to then be passed to the newly setup `_RegisterComponent` function, this would create a single point for all flavors to be able to support registering all visual elements.

**Comment by farfromrefug** [comment]
@NathanaelA i think registering only the use components is a subject of another rfc but totally doable today for flavors. They use a template compiler which can thus detect which components are.used. 
And just the whole don't add 8mb vs 30k for the code we are talking about. You are talking with source maps which should not be in counted ;)
Finally every little little gain become a big one when you add them up. No stone should be left unturned

**Comment by NathanaelA** [comment]
No source-maps in the TS.      Typo -- missed the `1.` somehow.     Should be `1.8mb` of basically source files lives in the Core/UI folder; of which only ~30k is the builder.    

Trying to extract out the builder seems to me to be a _lot_ more work as Builder class alone is referenced  ~70 times in different core source files and tied into places like the Modal, startup and of course navigation.   Which then means this code ALSO has to be abstracted out into the Vanilla flavor, and then every place that calls those have to be checked to see if abstracting it out  breaks something else that also would have to be abstracted out.   Its a domino effect...     In addition Builder is also referenced in ~200 plugins, that now have to be fixed to import the new Vanilla/Builder.   So extracting Builder into a Vanilla plugin seems like a **lot** of work for such a low payoff.   
  
Auto-registration idea could save a large chunk of that 1.8mb in a lot of apps, and seems like a lot lower hanging fruit with better bang for the buck and would also give the same wins to all flavors.

**Comment by farfromrefug** [comment]
@NathanaelA agree on auto registration being more important. But not the same rfc ! So it does not invalidate this one. 
Would you like to open a RFC for the auto registration?
